tools/doctest/src/doctest.ts

```diff
@@ -0,0 +1,331 @@
+  sourceMap.sourceRoot = '';
+  sourceMap.sources = [
+    relative(dirname(generatedFile), resolve(location.origin)),
+  ];
+  const consumer = await new SourceMapConsumer(sourceMap);
+  const generator = new SourceMapGenerator({
+    file: consumer.file,
+    sourceRoot: consumer.sourceRoot,
+  });
+  // We want descending order of the `generated` property.
+  const positions = [...location.positions].reverse();
+  consumer.eachMapping(mapping => {
+    // Note `mapping.originalLine` is the line number with respect to the
+    // extracted, raw code.
+    const {extracted, original} = positions.find(({extracted}) => {
+      return mapping.originalLine >= extracted;
+    })!;
+
+    // `original.line` will account for `extracted`, so we need to subtract
+    // `extracted` to avoid duplicity. We also subtract 1 because `extracted` is
+    // 1-indexed.
+    mapping.originalLine -= extracted - 1;
+
+    generator.addMapping({
+      ...mapping,
+      original: {
+        line: mapping.originalLine + original.line - 1,
+        column: mapping.originalColumn + original.column - 1,
+      },
+      generated: {
+        line: mapping.generatedLine,
+        column: mapping.generatedColumn,
+      },
+    });
+  });
+  return generator.toJSON();
+}
+
+async function compileCode(language: Language, code: string) {
+  return (await transform(code, {
+    sourceMaps: true,
+    inlineSourcesContent: false,
+    jsc: {
+      parser: {
+        syntax: (
+          {
+            [Language.TypeScript]: 'typescript',
+            [Language.JavaScript]: 'ecmascript',
+          } as const
+        )[language],
+      },
+      target: 'es2022',
+    },
+  })) as Required<Output>;
+}
+
+function* extractExampleCode(comments: Iterable<Comment>) {
+  interface Context {
+    language: Language;
+    fail: boolean;
+    start: number;
+  }
+  for (const {text, position: loc} of comments) {
+    const {tags} = parse(text, {
+      unwrap: true,
+      tags: ['example'],
+      lineNumbers: true,
+      preserveWhitespace: true,
+    });
+    for (const {description, lineNumber} of tags as Array<
+      Tag & {lineNumber: number}
+    >) {
+      if (!description) {
+        continue;
+      }
+      const lines = description.split('\n');
+      const blocks = [];
+      let context: Context | undefined;
+      for (let i = 0; i < lines.length; i++) {
+        const line = lines[i]!;
+        const borderIndex = line.indexOf(CODE_FENCE);
+        if (borderIndex !== -1) {
+          if (!context) {
+            const [tag, ...options] = line
+              .slice(borderIndex + CODE_FENCE.length)
+              .split(' ');
+            if (options.includes('ignore')) {
+              // Ignore the code sample.
+              continue;
+            }
+            const fail = options.includes('fail');
+            if (!tag || tag.match(/js|javascript/)) {
+              context = {language: Language.JavaScript, fail, start: i + 1};
```

Comment:
Add a comment on why the `i + 1` it seems hard to follow. (I presume it source map related)

Reply:
`i+1` is the starting line for the code block (nothing to do with the source map). I wouldn't say adding a comment would help here; if you get to this point, you need to know where the indexes line up.